audiostream2py/data.py

```diff
@@ -0,0 +1,148 @@
+"""AudioData dataclass for read values from PyAudioSourceReader.  Provides methods to concatenate
+multiple AudioData objs and slicing an individual AudioData obj while maintaining the correct
+timestamps and frame counts.
+"""
+
+import math
+from functools import cached_property
+from typing import Union, Sequence, Tuple
+from dataclasses import dataclass
+
+from audiostream2py.enum import PaStatusFlags
+
+
+@dataclass(frozen=True)
+class AudioData:
```

Comment:
Discussion (@andeaseme and @valentin-feron): Here it seems that AudioData is a container of a waveform segment containing its timestamp(s) as well as defining how these waveform segments should be concatenated. 

What are the pros/cons of this versus a more functional approach? 

Do we want to define `__add__`, and/or `__iadd__` and/or `extend` to point to "concatenate"?

Is `AudioSegment` perhaps a better name for this since we're talking about... audio segments (more precise)? Note that there's an object in `hear` called [AudioSegments](https://github.com/otosense/hear/blob/master/hear/audio_segments.py#L37), plural, so not clashing, but might want to have a look at it for some ideas on a related subject.

Reply:
> Do we want to define `__add__`, and/or `__iadd__` and/or `extend` to point to "concatenate"?

These three don't fit into our expected usage.  They imply that we have a starting block and we add to them incrementally.  Our use case is to query a range and receive all blocks at once.  

They would encourage unnecessary operations and memory usage. These audio blocks are expected to be short and numerous. Adding each of them, or concatenating two at a time, would create a new byte object for each operation along with extra arithmetic calculations.

> Is `AudioSegment` perhaps a better name for this...

I like `AudioSegment` but also `AudioBlock` to reinforce `block` as the nomenclature for `stream2py` buffer items